### ISSUE TITLE
Massive memory leak since 8.5.5 / leaking the entire stack frame and all local variables everywhere

### ISSUE THREAD

**Issue by luk2302** [lovasoa/marshmallow_dataclass#198]
In https://github.com/lovasoa/marshmallow_dataclass/pull/187 a change was implemented that included the current stack frame in an internal cache, therefore causing all local variables of the entire call stack to never be freed. Demo:

```python
from dataclasses import dataclass
import marshmallow_dataclass
from marshmallow.schema import BaseSchema
from marshmallow_dataclass import _internal_class_schema


class Outer:
    def __del__(self):
        print("deleting outer...")


@dataclass(frozen=True)
class Sample:
    a: str
    def __del__(self):
        print("deleting sample...")


def inner_demo():
    schema = marshmallow_dataclass.class_schema(Sample, base_schema=BaseSchema)()
    data = schema.load({"a": "a string"})
    return data


def demo():
    d = inner_demo()
    o = Outer()


for i in range(10):
    demo()
    print(_internal_class_schema.cache_info())

print("done with demo, no object deleted so far")
_internal_class_schema.cache_clear()
print("cleared cache")
```

No `__del__` is called on either `Sample` or `Outer` during the method invocations, only after the `cache_clear` can the local variables be garbage collected. If you run the demo with 8.4.4 or just replace `_internal_class_schema(clazz, base_schema, clazz_frame)` with `_internal_class_schema(clazz, base_schema, None)` in [marshmallow_dataclass.class_schema](https://github.com/lovasoa/marshmallow_dataclass/blob/master/marshmallow_dataclass/__init__.py#L356). Additionally this also demonstrates that the caching does no longer work as there no cache hits at all anymore.

**Comment by LostInDarkMath** [comment]
Hi,
I came across this issue because our software has a memory leak in production. After days of troubleshooting and narrowing down the problem, I came across this ticket.

Are there any plans to fix this in the near future?
And how can it be that such a critical bug remains unfixed for 2 years?

**Comment by lovasoa** [comment]
I understand your frustration, but please refrain from "how can it be", when referring to free and open source software. Instead, you can offer time or money to free software projects to unblock situations like this.

**Comment by dairiki** [comment]
This is (mostly) fixed by #258 (marshmallow_dataclass >= 8.6.1).

When necessary, references to frames are still held until the schema is constructed (by accessing the `.Schema` attribute). Memory leakage is still possible if large numbers of `marshmallow_dataclasses` are constructed without accessing their `.Schema` attribute.

**Comment by vitaly-krugl** [comment]
Hi @dairiki - I wanted to follow up regarding:
> Memory leakage is still possible if large numbers of marshmallow_dataclasses are constructed without accessing their .Schema attribute.

Does this mean that if a large number of these dataclasses are constructed using the dataclass constructor directly (without unmarshalling from data), then we still have a memory leak situation in the latest version of marshmallow-dataclass?

If so, what is the point of deferring schema construction, if deferring leads to serious issues? The deferring is also responsible for a terrible race condition in multi-threaded applications - see https://github.com/lovasoa/marshmallow_dataclass/issues/282. Can the `marshmallow_dataclass.dataclass` decorator construct the schema inline instead (without deferring), and put all of these issues to rest?

Also, in this context, I am assuming that calling `class_schema()` on the dataclass is equivalent to accessing its `.Schema` attribute. Is this correct?
 
If I have a marshmallow dataclass and schema generation defined at top level of module like below, do I still need to be concerned about memory leaks after your fix?
```
@marshmallow_dataclass.dataclass
class MyDataclass:
  s: str
  i: int

MyDataclassSchema = marshmallow_dataclass.class_schema(MyDataclass)()
```

**Comment by vkruglik-aka** [comment]
Hi @lovasoa, sorry to bother you - could you please chime in on my comment above - https://github.com/lovasoa/marshmallow_dataclass/issues/198#issuecomment-2849480715? Many thanks in advance.

**Comment by lovasoa** [comment]
Hi! I'm sorry, I don't use this package myself anymore and currently do not have a lot of bandwidth to maintain it. Happy to add new maintainers, though!

**Comment by vkruglik-aka** [comment]
Thank you for following up, @lovasoa. Has there been an attempt to get the core marshmallow project to take marshmallow-dataclass under its wing? They seem like a great pairing.

**Comment by dairiki** [comment]
> Does this mean that if a large number of these dataclasses are constructed using the dataclass constructor directly (without unmarshalling from data), then we still have a memory leak situation in the latest version of marshmallow-dataclass?

My memory is a little foggy on this, but in the "normal" case of dataclasses that are defined at the top level of a module, there is no longer a memory leak. The comment at the top of `_maybe_get_callers_frame` describes the circumstances where we need to hang onto a reference to the frame where the dataclass was constructed.

https://github.com/lovasoa/marshmallow_dataclass/blob/9812e50d3324d3e78d1094a6ed62417e2b03f987/marshmallow_dataclass/__init__.py#L97-L118



> If so, what is the point of deferring schema construction, if deferring leads to serious issues?

In general, we need to defer construction of the `.Schema` attribute in order to deal with forward type references.

E.g. the following code is legal, but `A.Schema` can not be computed at the time that `A` is defined, since the type of `B` is still unknown at that point.

```py
from __future__ import annotations

from marshmallow_dataclass import dataclass


@dataclass
class A:
   b: B

@dataclass
class B:
   x: int
```

In my own code, I avoid the use of `marshmallow_dataclass.dataclass` (in favor of vanilla `dataclasses.dataclass`), then use `marshmallow_dataclass.class_schema` to construct the schema explicitly.  This allows control of exactly when the schema class is generated.


```py
from __future__ import annotations

from dataclasses import dataclass

from marshmallow_dataclass import class_schema

@dataclass
class A:
   b: B

@dataclass
class B:
   x: int

# NB: Note that this would fail if executed before B is defined.
A_schema = class_schema(A)
```

**Comment by vkruglik-aka** [comment]
That's good to know, thank you very much @dairiki for the details as well as for the examples. By the way, regarding `from dataclasses import dataclass, class_schema`, I think `class_schema` needs to come from `marshmallow_dataclass` instead of `dataclasses`. Everything else looks good.

By the way, do you know if using `class_schema()` on a marshmallow dataclass is equivalent to `.Schema`? Empirically, `class_schema()` seems to work on a marshmallow dataclass, but I can't tell if there might be some hidden problems with using it this way.

**Comment by dairiki** [comment]
> regarding `from dataclasses import dataclass, class_schema`, I think `class_schema` needs to come from `marshmallow_dataclass` instead of `dataclasses`

Doh.  Of course, thank you.  I've edited my comment to fix that.



> By the way, do you know if using `class_schema()` on a marshmallow dataclass is equivalent to `.Schema`? Empirically, `class_schema()` seems to work on a marshmallow dataclass, but I can't tell if there might be some hidden problems with using it this way.

Yes, `.Schema` (on first evaluation) just calls `class_schema` on the dataclass (passing any `base_schema` passed to the dataclass decorator).  (The code is not the most readable, but that happens [here](https://github.com/lovasoa/marshmallow_dataclass/blob/9812e50d3324d3e78d1094a6ed62417e2b03f987/marshmallow_dataclass/__init__.py#L277-L281)).
Though doing that seems kind of confusing and pointless, and — unless the `.Schema` attribute is accessed separately — I think it will leave the deferred `.Schema` attribute unevaluated (with the possibility of memory leakage).

**Comment by vkruglik-aka** [comment]
Thanks @dairiki. And regarding:
> unless the .Schema attribute is accessed separately — I think it will leave the deferred .Schema attribute unevaluated (with the possibility of memory leakage).

I think you are saying that if we use the `@dataclasses.dataclass` decorator to define our dataclasses at top level (instead of `@marshmallow_dataclass.dataclass`), then we don't have to worry about the memory leak, because there would be no deferred `.Schema` attribute that needs to be used.

**Comment by dairiki** [comment]
> if we use the @dataclasses.dataclass decorator to define our dataclasses at top level (instead of @marshmallow_dataclass.dataclass), then we don't have to worry about the memory leak, because there would be no deferred .Schema attribute that needs to be used.

Yes. Exactly!

**Comment by vkruglik-aka** [comment]
Just found out that `marshamllow_dataclass.dataclass` is also required for this use case: 
```
@marshamllow_dataclass.dataclass(base_schema=SomeMarshmallowSchema)
class ...
```

cc @dairiki

**Comment by dairiki** [comment]
> Just found out that `marshamllow_dataclass.dataclass` is also required for this use case:
> 
> ```
> @marshamllow_dataclass.dataclass(base_schema=SomeMarshmallowSchema)
> class ...
> ```

One may pass the `base_schema` parameter directly to  `class_schema`.  (That's all `marshmallow_dataclass.dataclass` does — it arranges for the `.Schema` attribute to be constructed by calling `class_schema` with that parameter.)

```py
MyclassSchema = marshmallow_dataclass.class_schema(MyClass, base_schema=SomeMarshmallowSchema)
```

Documentation [here](https://github.com/lovasoa/marshmallow_dataclass/tree/master?tab=readme-ov-file#custom-mapping-between-classes-and-fields).

(The `base_schema` specifies a base class for the constructed Marshmallow schema, not a base class for the dataclass.  This is more clear when calling `class_schema` — which inspects the passed dataclass and constructs a Marshmallow schema to match — directly.)

----

Had it been up to me, I would never have included `marshmallow_dataclass.dataclass` in our API.  It might be a little "simpler" to use (than calling `class_schema`) in basic cases, but it obscures the workings so that once the use case gets complex enough that it matters, it's harder to understand how things actually work (and therefore how to fix things).

In any case, I am fairly certain that there is nothing that `marshmallow_dataclass.dataclass` can do that isn't already possible using `class_schema`.
